src/query/point/point_composite_shape/query_options_dispatcher.rs

```diff
@@ -0,0 +1,96 @@
+use alloc::boxed::Box;
+use core::any::{Any, TypeId};
+
+use crate::query::gjk::GjkOptions;
+use crate::query::point::point_query::QueryOptions;
+use crate::shape::Compound;
+use hashbrown::HashMap;
+
+/// Retrieves a stored option which should be used for given shape.
+pub trait QueryOptionsDispatcher {
+    /// Retrieves a stored option which should be used for given shape.
+    fn get_option_for_shape(&self, shape_type_id: &TypeId) -> &dyn QueryOptions;
+}
+
+impl QueryOptionsDispatcher for () {
+    fn get_option_for_shape(&self, _shape_type_id: &TypeId) -> &dyn QueryOptions {
+        &()
+    }
+}
+
+/// Implements [QueryOptionsDispatcher] with a generic dispatch of options.
+/// This supports Compound shapes, user-defined shapes and user-defined algorithms.
+pub struct QueryOptionsDispatcherMap {
+    /// Contains options to apply for shapes query algorithms.
+    ///
+    /// - Because algorithms can be recursive, Rc is used.
+    /// - Because algorithms can contain user-defined shapes, TypeId is used.
+    /// - Because algorithms can contain user-defined algorithms (and options), `Box<dyn>` is used.
+    pub options_for_shape: HashMap<TypeId, Box<fn(&Self) -> &dyn QueryOptions>>,
+
+    /// Store options inside here.
+    /// It's used to capture variables, as [Self::options_for_shape] should be cloneable.
+    pub options: HashMap<TypeId, Box<dyn QueryOptions>>,
+}
+
+impl Default for QueryOptionsDispatcherMap {
+    fn default() -> QueryOptionsDispatcherMap {
+        let self_ref: Box<fn(&Self) -> &dyn QueryOptions> =
+            Box::new(|s: &QueryOptionsDispatcherMap| -> &dyn QueryOptions { s });
+        let gjk_options: Box<fn(&Self) -> &dyn QueryOptions> =
+            Box::new(|s: &QueryOptionsDispatcherMap| -> &dyn QueryOptions {
+                s.options[&TypeId::of::<GjkOptions>()].as_ref()
+            });
+        QueryOptionsDispatcherMap {
+            options_for_shape: [
+                (TypeId::of::<Compound>(), self_ref),
+                #[cfg(feature = "dim2")]
+                (
+                    TypeId::of::<crate::shape::ConvexPolygon>(),
+                    gjk_options.clone(),
+                ),
+                #[cfg(feature = "dim3")]
+                (TypeId::of::<crate::shape::ConvexPolyhedron>(), gjk_options),
```

Comment:
Do roundshapes and other container shapes need to be specific ? like `TypeId::of::<RolunShape<crate::shape::ConvexPolyhedron>>()` ?

Reply:
roundshapes use support point algorithm which is gjk options.